Helper/Api/Products.php

```diff
@@ -1053,4 +868,341 @@ public function cleanProductStorekeeperId(string $storeId): void
+                    );
+                }
+            }
+        }
+
+        $galleryImages = $target->getMediaGalleryImages()->getItems();
+        $existingImagesArray = [];
+        foreach ($galleryImages as $image) {
+            $imagePath = explode('/', parse_url($image->getFile(), PHP_URL_PATH));
+            $imageName = pathinfo(end($imagePath), PATHINFO_FILENAME);
+            $existingImagesArray[] = $imageName;
+        }
+
+        if (isset($flat_product['product_images'])) {
+            $mainImage = explode('/', $flat_product['main_image']['big_url']);
+            $mainImageName = pathinfo(end($mainImage), PATHINFO_FILENAME);
+            foreach ($flat_product['product_images'] as $product_image) {
+                $newImagePath = explode('/', parse_url($product_image['big_url'], PHP_URL_PATH));
+                $newImageName = pathinfo(end($newImagePath), PATHINFO_FILENAME);
+
+                $countDuplicates = count(preg_grep("/^{$newImageName}\_[0-9]+/", $existingImagesArray));
+
+                if ($newImageName
+                    !== $mainImageName
+                    && !preg_match("/^{$newImageName}\_[0-9]+/", $mainImageName)) {
+                    if (!in_array($newImageName, $existingImagesArray) && $countDuplicates <= 0) {
+                        $shouldUpdate = $this->setGalleryImage($product_image['big_url'], $target, false);
+                    }
+                }
+            }
+        }
+
+        return $shouldUpdate;
+    }
+
+    /**
+     * @param Product $target
+     * @param array $result
+     * @return void
+     * @throws LocalizedException
+     */
+    private function assignCategories(Product $target, array $result): void
+    {
+        // assign the store to the first available, otherwise delete operations will go wrong
+        $categoryIds = $this->getResultCategoryIds($result);
+        if (!empty($categoryIds)) {
+            // check if categories exist
+            if ($categories = $this->categoriesExist($categoryIds)) {
+                $categoryIds = array_map(function ($category) {
+                    return $category->getId();
+                }, $categories);
+
+                try {
+                    $diff = array_diff($target->getCategoryIds(), $categoryIds);
+                    if (empty($target->getCategoryIds()) || $diff) {
+                        if ($diff) {
+                            foreach ($diff as $categoryId) {
+                                $this->categoryLinkRepository->deleteByIds($categoryId, $target->getSku());
+                            }
+                        }
+
+                        $this->categoryLinkManagement->assignProductToCategories(
+                            $target->getSku(),
+                            $categoryIds
+                        );
+                    }
+                } catch (\Exception $e) {
+                    $this->logger->error($e->getMessage(), $this->logger->buildReportData($e));
+                }
```

Comment:
```suggestion
```

We should let it fails  category assingeing fails, otherwise we think the sync was succesfull